File: src/main/java/core/scm/specialforms/Lambda.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.exceptions.IllegalSyntaxException;
import core.scm.Cons;
import core.scm.Procedure;
import core.scm.Symbol;

import java.util.HashSet;
import java.util.List;

/* Syntax:
 * (lambda <formals> <body>)
 *
 * <formals>:
 * (<variable1> ...)
 * <variable>
 * (<variable1> ... <variablen> . <variablen+1>)
 */
public enum Lambda implements ISpecialForm {
  LAMBDA;

  @Override
  public Procedure eval(List<Object> expression, Environment env, Evaluator evaluator) {
    if (expression.size() < 3) {
      throw IllegalSyntaxException.of(toString(), expression);
    }

    List<Symbol> params;
    boolean variadic = false;
    /* Check if args is a List or not */
    Object args = expression.get(1);
    if (args instanceof List) {
      /* Check args for duplicates */
      if (!((List) args).isEmpty()) {
        HashSet<Object> temp = new HashSet<>(((List) args).size());
        for (Object o : ((List) args)) {
          if (!(o instanceof Symbol) && !(Cons.isPair(o))) {
            throw IllegalSyntaxException.of(toString(), expression, String.format("not an identifier: %s", o));
          }
          if (temp.contains(o)) {
            throw IllegalSyntaxException.of(toString(), expression, String.format("duplicate argument name: %s", o));
          }
          temp.add(o);
        }
      }
      /* (lambda (arg-id ...+) body ...+) OR
       * (lambda (arg-id ...+ . rest-id) body ...+) */
      if (Cons.isList(args)) {
        /* args is a proper list, hence non-variadic lambda */
        params = (List<Symbol>)args;
      } else {
        /* args is an improper list, hence variadic lambda */
        params = Cons.flatten((List<Symbol>)args);
        variadic = true;
      }
    } else {
      /* Variadic arity */
      /* (lambda rest-id body ...+) */
      if (!(args instanceof Symbol)) {
        throw new IllegalSyntaxException(String.format("lambda: bad argument sequence (%s) in form: %s",  args, expression));
      }
      params = Cons.list((Symbol)args);
      variadic = true;
    }
    Object body;
    if (expression.size() == 3) {
      body = expression.get(2);
    } else {
      /* Add implicit `begin` */
      body = Cons.list(Begin.BEGIN);
      ((List)body).addAll(expression.subList(2, expression.size()));
    }
    return new Procedure("", params, body, env, variadic);
  }

  @Override
  public String toString() {
    return "lambda";
  }
}

File: src/main/java/core/scm/specialforms/Lambda.kt
package core.scm.specialforms

import core.Evaluator
import core.environment.Environment
import core.exceptions.IllegalSyntaxException
import core.procedures.predicates.Predicate
import core.scm.Closure
import core.scm.Symbol
import core.Writer
import core.procedures.Arity.AtLeast
import core.procedures.Arity.Exactly
import core.utils.Utils
import java.util.LinkedList

/**
 * Syntax:
 * (lambda <formals> <body>)
 *
 * <formals>:
 * (<variable1> ...)
 * <variable>
 * (<variable1> ... <variableN> . <variableN+1>)
 */
object Lambda : SpecialForm("lambda") {

    override fun eval(form: List<Any?>, env: Environment, evaluator: Evaluator): Closure {
        if (form.size < 3) {
            throw IllegalSyntaxException(toString(), Writer.write(form))
        }
        /* Check if args is a List or not */
        val args = form[1]
        val variadic = args is Symbol || !Predicate.isProperList(args)
        val params = when (args) {
            is List<*>, is Pair<*, *> -> when (variadic) {
                true  -> validateParamsList(flatten(Utils.toSequence(args)).asSequence(), form)
                false -> validateParamsList(Utils.toSequence(args), form)
            }
            /* Variadic arity */
            /* (lambda rest-id body ...+) */
            is Symbol -> listOf(args)
            else -> throw IllegalSyntaxException(toString(), Writer.write(form), "bad argument sequence: ($args)")
        }
        /* Get the body and add implicit Begin if multiple body forms */
        val body  = if (form.size == 3) form[2] else listOf(Begin) + form.drop(2)
        val arity = if (variadic) AtLeast(0) else Exactly(params.size)
        return Closure(params, body, env, arity)
    }

    /* Validate args */
    private fun validateParamsList(seq: Sequence<*>, form: List<Any?>): List<Symbol?> {
        if (seq.any()) {
            val temp = hashSetOf<Symbol>()
            seq.forEach {
                when {
                    it !is Symbol -> throw IllegalSyntaxException(toString(), Writer.write(form), "not an identifier: ${Writer.write(it)}")
                    !temp.add(it) -> throw IllegalSyntaxException(toString(), Writer.write(form), "duplicate argument name: $it")
                }
            }
        }
        @Suppress("UNCHECKED_CAST")
        return (seq as Sequence<Symbol?>).toList()
    }

    /* Non-recursively flatten a list (or a chain of conses) */
    private fun flatten(seq: Sequence<*>) = mutableListOf<Any?>().apply {
        val queue = LinkedList(seq.toList())
        while (!queue.isEmpty()) {
            when (val e = queue.remove()) {
                is Pair<*, *> -> queue.apply {
                    add(e.first)
                    add(e.second)
                }
                is List<*> -> queue.addAll(e)
                else -> add(e)
            }
        }
    }
}
